src/Nethermind/Nethermind.BeaconNode.Storage/MemoryStore.cs

```diff
@@ -16,75 +16,211 @@
-            IDictionary<Checkpoint, BeaconState> checkpointStates,
-            IDictionary<ValidatorIndex, LatestMessage> latestMessages,
-            ILogger<MemoryStore> logger,
-            IOptionsMonitor<TimeParameters> timeParameterOptions)
+        private const string InMemoryDirectory = "memorystore";
+
+        public MemoryStore(ILogger<MemoryStore> logger,
+            IOptionsMonitor<InMemoryConfiguration> inMemoryConfigurationOptions,
+            DataDirectory dataDirectory,
+            IFileSystem fileSystem)
+        {
+            _logger = logger;
+            _inMemoryConfigurationOptions = inMemoryConfigurationOptions;
+            _dataDirectory = dataDirectory;
+            _fileSystem = fileSystem;
+            _jsonSerializerOptions = new JsonSerializerOptions {WriteIndented = true};
+            _jsonSerializerOptions.ConfigureNethermindCore2();
+            if (_inMemoryConfigurationOptions.CurrentValue.LogBlockJson ||
+                inMemoryConfigurationOptions.CurrentValue.LogBlockStateJson)
+            {
+                _ = GetLogDirectory();
+            }
+        }
+
+        public Checkpoint BestJustifiedCheckpoint { get; private set; } = Checkpoint.Zero;
+        public Checkpoint FinalizedCheckpoint { get; private set; } = Checkpoint.Zero;
+        public ulong GenesisTime { get; private set; }
+        public bool IsInitialized { get; private set; }
+        public Checkpoint JustifiedCheckpoint { get; private set; } = Checkpoint.Zero;
+        public ulong Time { get; private set; }
+
+        public ValueTask<BeaconBlock> GetBlockAsync(Root blockRoot)
+        {
+            if (!_blocks.TryGetValue(blockRoot, out BeaconBlock? beaconBlock))
+            {
+                throw new ArgumentOutOfRangeException(nameof(blockRoot), blockRoot, "Block not found in store.");
+            }
+
+            return new ValueTask<BeaconBlock>(beaconBlock!);
+        }
+
+        public ValueTask<BeaconState> GetBlockStateAsync(Root blockRoot)
+        {
+            if (!_blockStates.TryGetValue(blockRoot, out BeaconState? state))
+            {
+                throw new ArgumentOutOfRangeException(nameof(blockRoot), blockRoot, "State not found in store.");
+            }
+
+            return new ValueTask<BeaconState>(state!);
+        }
+
+        public ValueTask<BeaconState?> GetCheckpointStateAsync(Checkpoint checkpoint, bool throwIfMissing)
+        {
+            if (!_checkpointStates.TryGetValue(checkpoint, out BeaconState? state))
+            {
+                if (throwIfMissing)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(checkpoint), checkpoint,
+                        "Checkpoint state not found in store.");
+                }
+            }
+
+            return new ValueTask<BeaconState?>(state);
+        }
+
+        public async IAsyncEnumerable<Root> GetChildKeysAfterSlotAsync(Root parent, Slot slot)
+        {
```

Comment:
I assume that this is just a memory store that does not have to be super fast as it is used for testing

Reply:
Yes. The interface is written with async support, assuming that it will be doing database like operations at some point and actually have to fetch stuff from disk (I haven't looked at the Eth1 implementation yet).

But currently it is all in-memory, with simple logging added to dump (one way) to disk for analysis and debugging.